### PR TITLE
Add orka-full task for MacStadium Orka integration

### PR DESCRIPTION
**Pull request by spikeburton**
# Changes

This PR aims to add a Task in order to integrate Tekton with [Orka by MacStadium](https://www.macstadium.com/orka). Orka allows users to run macOS builds and workloads on real Apple hardware. An Orka / Tekton integration will allow the utilization of macOS build agents for CI jobs that run on Tekton pipelines.

The `orka-full` Task will allow the utilization of a single macOS build agent, allowing the user to provide a build script as a parameter, and will allow the user to store build artifacts in a Tekton workspace. 

> **NOTE**: In order to use this Task, an Orka environment is required. Orka runs in a self-contained Kubernetes cluster and a VPN connection must be established between the Orka cluster and the Kubernetes cluster in which Tekton is installed.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

File: task/orka-full/0.1/orka-full.yaml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+kind: Task
+metadata:
+  name: orka-full
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.16.0"
+spec:
+  description: >-
+    With this set of Tasks, you can use your Orka environment
```
Comment (sbwsg) [inline task/orka-full/0.1/orka-full.yaml:12]: nit: `With this set of Tasks` -> `With this Task`?

**Review by sbwsg — COMMENTED** [review]
I wonder if it might make sense to include a **Manual setup** section for Installation in the READMEs that calls out the `orka-tekton-config` ConfigMap .  If a user discovers and installs these Tasks from the Hub they may miss the fact that they also need to setup the Orka API ConfigMap?

**Comment by spikeburton** [comment]
> I wonder if it might make sense to include a **Manual setup** section for Installation in the READMEs that calls out the `orka-tekton-config` ConfigMap .  If a user discovers and installs these Tasks from the Hub they may miss the fact that they also need to setup the Orka API ConfigMap?

@ikoevska do you think we can document this? Seems like a good idea to me 🙂

File: task/orka-full/0.1/orka-full.yaml (hunk 1 above)
Comment (spikeburton) [reply task/orka-full/0.1/orka-full.yaml:12]: Good catch

**Review by vinamra28 — CHANGES_REQUESTED** [review]
just a few concerns that needs to be addressed :slightly_smiling_face: 
/hold

File: task/orka-modular/0.1/orka-modular.yaml [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+          orka-cleanup
+          exit 1
+        fi
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: orka-deploy
+  labels:
+    app.kubernetes.io/version: "0.1"
```
[the remark this answers is not in the document]
Comment (sbwsg) [reply task/orka-modular/0.1/orka-modular.yaml:127]: Hm, for reference, I came away from the October 1 Catalog / Hub WG with the impression that this was the suggested approach to group closely-related Tasks for now.  Created issue https://github.com/tektoncd/catalog/issues/540 to discuss further.

**Comment by sbwsg** [comment]
@spikeburton in order to trigger EasyCLA you may need to force-push an empty change (add an extra newline to commit message or similar) or try closing and re-opening the PR.  Sorry about that 😓

File: task/orka-modular/0.1/orka-modular.yaml [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        The name of the config map, which stores the name of the generated VM configuration.
+      default: orka-vm-name
+    - name: orka-vm-name-config-key
+      type: string
+      description: |
+        The name of the key in the VM name config map, which stores the name of the generated VM configuration.
+      default: vm-name
+  steps:
+    - name: copy-script
+      image: macstadium/orka-tekton-runner:2020-10-23-a93757dc-0.1
```
Comment (spikeburton) [inline task/orka-modular/0.1/orka-modular.yaml:211]:
```suggestion
      image: docker.io/macstadium/orka-tekton-runner:2020-10-23-a93757dc-0.1@sha256:e8ed3370dcb91cdb8bcd4e9a7e9f6652879d80acdab9644cda69a98cd8c93339
```

File: task/orka-modular/0.1/orka-modular.yaml [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+      script: |
+        #!/bin/sh
+        SCRIPT=$(mktemp)
+        # Safeguard against having to escape quotes / vars in script
+        cat > $SCRIPT << 'EOF'
+        $(params.script)
+        EOF
+        copy-script $SCRIPT
+    - name: build
+      image: macstadium/orka-tekton-runner:2020-10-23-a93757dc-0.1
```
Comment (spikeburton) [inline task/orka-modular/0.1/orka-modular.yaml:221]:
```suggestion
      image: docker.io/macstadium/orka-tekton-runner:2020-10-23-a93757dc-0.1@sha256:e8ed3370dcb91cdb8bcd4e9a7e9f6652879d80acdab9644cda69a98cd8c93339
```

File: task/orka-modular/0.1/orka-modular.yaml [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        The name of the config map, which stores the name of the generated VM configuration.
+      default: orka-vm-name
+    - name: orka-vm-name-config-key
+      type: string
+      description: |
+        The name of the key in the VM name config map, which stores the name of the generated VM configuration.
+      default: vm-name
+  steps:
+    - name: teardown
+      image: macstadium/orka-tekton-runner:2020-10-23-a93757dc-0.1
```
Comment (spikeburton) [inline task/orka-modular/0.1/orka-modular.yaml:310]:
```suggestion
      image: docker.io/macstadium/orka-tekton-runner:2020-10-23-a93757dc-0.1@sha256:e8ed3370dcb91cdb8bcd4e9a7e9f6652879d80acdab9644cda69a98cd8c93339
```

**Comment by spikeburton** [comment]
Hmm, not sure why the integration tests are failing now. All tests pass when run locally after rebase.

File: task/orka-full/0.1/README.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+## Installation
+
+Before you can use this `Task` in Tekton pipelines, you need to install it and the Orka configuration in your Kubernetes cluster.
+
+**Default namespace installation**
+
+To install in your Kubernetes cluster's `default` namespace, run the following command against your actual Orka API endpoint.
+
+```sh
+ORKA_API=http://10.221.188.100 ./install.sh --apply
```
[the remark this answers is not in the document]
Comment (spikeburton) [reply task/orka-full/0.1/README.md:36]: @ikoevska I've modified the install scripts as requested, so we will need to update the documentation here for how to fetch each script mentioned. Maybe we can just add a note where each script is mentioned, something like:

> **NOTE**: if installing from Tekton hub, you can get the install script by running the command `curl -LO https://raw.githubusercontent.com/tektoncd/catalog/master/task/orka-full/0.1/install.sh && chmod 755 install.sh`

File: task/orka-full/0.1/add-orka-creds.sh [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+if [[ ! $EMAIL || ! $PASSWORD ]]; then
+  echo -e "Email and password required!\n"
+  echo "$USAGE"
+  exit 1
+fi
+
+sed -e 's|$(email)|'"$EMAIL"'|' \
+  -e 's|$(password)|'"$PASSWORD"'|' resources/orka-creds.yaml.tmpl \
+  > resources/orka-creds.yaml
+kubectl $ACTION --namespace=$NAMESPACE -f resources/orka-creds.yaml
```
[the remark this answers is not in the document]
Comment (spikeburton) [reply task/orka-full/0.1/add-orka-creds.sh]: Have updated all scripts to use absolute paths.

File: task/orka-modular/0.1/orka-modular.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (spikeburton) [reply task/orka-modular/0.1/orka-modular.yaml:127]: Am going to hold on any changes here for the time being - let me know what the internal consensus is and I will adapt accordingly

**Comment by chmouel** [comment]
Sounds like there is an issue with the testrunner, let me have a look.

(and it would be great if you can rebase your commits or at least between two tasks of orka-full or orka-medium)

**Comment by chmouel** [comment]
testing fix in #568 556

**Comment by chmouel** [comment]
Fix is properly tested in https://prow.tekton.dev/log?job=pull-tekton-catalog-integration-tests&id=1321840097338855424, you can rebase against origin/master when #569 would be merged.

Not that I did run your tests against a cluster with restricted RBAC (i.e: openshift) and your tests is failing, not an issue here as we don't have that policy, but that something you may want to be aware, if it's an absolute requirement just ask for the priveleged securityContext in your task.

**Comment by chmouel** [comment]
Any chances to rebase the commits please? And please look at this for commit messages : 

https://github.com/tektoncd/community/blob/master/standards.md?rgh-link-date=2020-10-23T21%3A50%3A35Z#commit-messages

**Comment by chmouel** [comment]
If you are ready for review you can remove the hold with the command "/hold cancel" (at the beginning of a comment)

**Comment by spikeburton** [comment]
@chmouel we still need to address some documentation changes when our tech writer has a chance 🙂

There also appears to be some disagreement in how to submit tightly coupled Tasks (as with `orka-modular` here) - I am unsure how to proceed until that is resolved.

File: task/orka-modular/0.1/orka-modular.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (spikeburton) [reply task/orka-modular/0.1/orka-modular.yaml:127]: @vinamra28 @sbwsg is there any additional feedback on how to proceed here?

[the remark this answers is not in the document]
Comment (sbwsg) [reply task/orka-modular/0.1/orka-modular.yaml:127]: I think it's best to move ahead following @vinamra28's direction here.  We can discuss further via #540 and probably in a follow-up TEP, but right now it sounds like there are technical reasons to stick with one catalog Task per directory.

If you prefer to hold off on doing this another approach would be to reduce this PR to just the orka-full task and then make a separate PR with the modular Tasks, which we can then use the case study for a TEP.

**Comment by chmouel** [comment]
No problems, let me know if I can be any help

**Comment by spikeburton** [comment]
/hold cancel

Ready for review :)

**Comment by spikeburton** [comment]
@chmouel @sbwsg @vinamra28 

Bump 🙂 ready for review

File: task/orka-full/0.1/README.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+kind: ConfigMap
+metadata:
+  name: orka-tekton-config
+data:
+  ORKA_API: http://10.221.188.100
+```
+
+```sh
+kubectl apply --namespace=<namespace> -f orka-configuration.yaml
+kubectl apply --namespace=<namespace> -f https://raw.githubusercontent.com/tektoncd/catalog/orka-0.1/task/orka-full/0.1/orka-full.yaml
```
Comment (vinamra28) [inline task/orka-full/0.1/README.md:43]:
```suggestion
kubectl apply --namespace=<namespace> -f https://raw.githubusercontent.com/tektoncd/catalog/task/orka-full/0.1/orka-full.yaml
```

**Review by vinamra28 — CHANGES_REQUESTED** [review]
just one small fix otherwise LGTM

**Review by vinamra28 — APPROVED** [review]
can you please squash the commits
(probably you can do once others have provided their comments) :slightly_smiling_face: 
/cc @chmouel @sbwsg @vdemeester

**Comment by sbwsg** [comment]
Awesome!

/lgtm

**Pull request merged**